internal/setting/handler_test.go

```diff
@@ -0,0 +1,167 @@
+	"context"
+	"encoding/json"
+	"github.com/NYCU-SDC/clustron-backend/internal/jwt"
+	"github.com/NYCU-SDC/clustron-backend/internal/setting"
+	"github.com/NYCU-SDC/clustron-backend/internal/setting/mocks"
+	"github.com/go-playground/validator/v10"
+	"github.com/google/uuid"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/mock"
+	"go.uber.org/zap"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+)
+
+var exampleValidKey = "ssh-rsa AAAAB3NzaC1yc2EAAAABJQAAAQB/nAmOjTmezNUDKYvEeIRf2YnwM9/uUG1d0BYsc8/tRtx+RGi7N2lUbp728MXGwdnL9od4cItzky/zVdLZE2cycOa18xBK9cOWmcKS0A8FYBxEQWJ/q9YVUgZbFKfYGaGQxsER+A0w/fX8ALuk78ktP31K69LcQgxIsl7rNzxsoOQKJ/CIxOGMMxczYTiEoLvQhapFQMs3FL96didKr/QbrfB1WT6s3838SEaXfgZvLef1YB2xmfhbT9OXFE3FXvh2UPBfN+ffE7iiayQf/2XR+8j4N4bW30DiPtOQLGUrH1y5X/rpNZNlWW2+jGIxqZtgWg7lTy3mXy5x836Sj/6L"
+
+func TestHandler_AddUserPublicKeyHandler(t *testing.T) {
+	testCase := []struct {
+		name           string
+		body           setting.AddPublicKeyRequest
+		expectedStatus int
+		expectError    bool
+	}{
+		{
+			name: "Should add public key",
+			body: setting.AddPublicKeyRequest{
+				Title:     "Test Title",
+				PublicKey: exampleValidKey,
+			},
+			expectedStatus: http.StatusOK,
+		},
+		{
+			name: "Should return error when public key is invalid",
+			body: setting.AddPublicKeyRequest{
+				Title:     "Test Title",
+				PublicKey: "MFswDQYJKoZIhvcNAQEBBQADSgAwR/",
+			},
+			expectedStatus: http.StatusBadRequest,
+		},
+		{
+			name: "Should return error when title is empty",
+			body: setting.AddPublicKeyRequest{
+				PublicKey: exampleValidKey,
+			},
+			expectedStatus: http.StatusBadRequest,
+		},
+		{
+			name: "Should return error when public key is empty",
+			body: setting.AddPublicKeyRequest{
+				Title: "Test Title",
+			},
+			expectedStatus: http.StatusBadRequest,
+		},
+	}
+
+	// Mock the dependencies
+	logger, err := zap.NewDevelopment()
+	if err != nil {
+		t.Fatalf("failed to create logger: %v", err)
+	}
+	store := mocks.NewStore(t)
+	store.On("AddPublicKey", mock.Anything, setting.AddPublicKeyParams{
+		UserID:    uuid.MustParse("7942c917-4770-43c1-a56a-952186b9970e"),
+		Title:     "Test Title",
+		PublicKey: exampleValidKey,
+	}).Return(setting.PublicKey{
+		ID:        uuid.MustParse("33a40641-45bb-4b47-aa33-113c7c562328"),
+		UserID:    uuid.MustParse("7942c917-4770-43c1-a56a-952186b9970e"),
+		Title:     "Test Title",
+		PublicKey: exampleValidKey,
+	}, nil)
+
+	h := setting.NewHandler(validator.New(), logger, store)
+
+	for _, tc := range testCase {
+		t.Run(tc.name, func(t *testing.T) {
+			requestBody, err := json.Marshal(tc.body)
+			if err != nil {
+				t.Fatalf("failed to marshal request body: %v", err)
+			}
+			r := httptest.NewRequest(http.MethodPost, "/api/setting/publicKey", bytes.NewReader(requestBody))
+			r = r.WithContext(context.WithValue(r.Context(), jwt.UserContextKey, jwt.User{
+				ID:       "7942c917-4770-43c1-a56a-952186b9970e",
+				Username: "testuser",
+				Role:     "user",
+			}))
```

Comment:
Consider adding a line break here to better separate logical sections of the code (prepare request -> send request)